Switch to Go 1.24 as a min version, bump CI, modernize sources

kolyshkin opened the pull request:
See individual commits for details. High level overview:

 - require go 1.24, add go 1.25 to CI, drop go 1.23
 - replace `omitempty` with `omitzero` everywhere (mostly this is a no-op and done for consistency, but for some fields, like nested struct fields, it will result in more "omitting" behavior)
 - add some golangci-extra exceptions for names (and moved some more from code to config)
 - modernize code for Go 1.24
 
 Similar runc PR: https://github.com/opencontainers/runc/pull/4851

kolyshkin committed:
Bump go to 1.24

Bump go to 1.24 in go.mod. Remove go 1.23 and add go 1.25 to CI.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
Modernize the code for Go 1.24

Mostly done by

	modernize -fix -test ./...

with some minor manual edits on top.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
stats.go: replace omitempty with omitzero

For some fields (like nested struct fields), omitempty has no effect
but omitzero does. Let's use it.

NOTE it might result in some compatibility issues since now
zero-only fields are not serialized.

Since linter-extra thinks it's a new code, add some naming exceptions.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Rebased.

kolyshkin commented on stats.go:
```diff
@@ -48,26 +48,26 @@ type PSIData struct {
-	ThrottlingData ThrottlingData `json:"throttling_data,omitempty"`
-	PSI            *PSIStats      `json:"psi,omitempty"`
-	BurstData      BurstData      `json:"burst_data,omitempty"`
+	CpuUsage       CpuUsage       `json:"cpu_usage,omitzero"`
```
The third option is to ignore linters-extra warnings as they are technically for new code and this is not new code. In the next commit or PR linters-extra won't flag this issue as it won't be "new".

kolyshkin replied:
I'm keeping the code as is for now, so you can take a look @lifubang

cyphar replied:
I agree with keeping it as-is. The mess with lint-extra is unfortunate but such is the cost of working on "legacy" code. ;)

kolyshkin replied:
Ah! Ended up naming exception to golangi-extra config (hoping we won't add new IDs with the whitelisted names).

kolyshkin replied:
...and moving existing exceptions to config file, too -- yay to less code clutter!

kolyshkin committed:
config_linux.go: replace omitempty with omitzero

For some fields (like nested struct fields), omitempty has no effect
but omitzero does. Let's use it (everywhere -- for consistency).

NOTE it might result in some compatibility issues since now
zero-only fields are not serialized.

Since linter-extra thinks it's a new code, add some naming exceptions.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
devices/config: replace omitempty with omitzero

This is technically a no-op, but let's change it for consistency.

While at it, move linter-extra exceptions to golangci-extra.yml.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@opencontainers/cgroups-maintainers can we have a second approval here?
